### PR TITLE
feat: Support raw github urls

### LINKED ISSUE #80: Support https://raw.githubusercontent.com
## Summary

What:

Alongside supporting deno.land and x.nest.land, add support for https://raw.githubusercontent.

Why:

So dmm can update dependencies imported from GitHub.

## Acceptance Criteria

Below is a list of tasks that must be completed before this issue can be closed.

- [ ] Write documentation (https://github.com/drashland/website/issues/225)
- [ ] Write unit tests
- [ ] Develop the feature

## Taking on this issue? Read this before you get started.

If you take on this issue and the acceptance criteria does not make sense, post comments on this issue with your questions. Also, you can always hop in the Discord channel at https://discord.gg/c5R3sYuR to get help quickly.

Please read the [Contributing Guidelines](https://github.com/drashland/.github/blob/master/CONTRIBUTING.md) before starting this issue.

### PR DESCRIPTION
**Pull request by ebebbington**
Fixes #80 

## Summary

- [x] Adds support for raw github urls (only versioned of course)
- [x] Approval 1
- [x] Approval 2

### CODE REVIEW

**Comment by ebebbington** [comment]
MMMMMmmmmmmm checks pass first time, that. is. what. i'm. talking. about

File: src/services/github_service.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  locked: boolean | null;
+  malicious: boolean | null;
+  unlisted: boolean;
+  updatedAt: string;
+  createdAt: string;
+}
+
+export default class GitHubService {
+  /**
+   * Fetches the latest release of a a github repository
```
Comment (crookse) [inline src/services/github_service.ts:19]: fetches ... of a a a a a a a a a a a a a a a

File: src/services/github_service.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   *
+   * @param githubUrl - Url of the github repository, eg https://github.com/drashland/dmm
+   *
+   * @returns The latest version.
```
Comment (crookse) [inline src/services/github_service.ts:23]: we going with @return or @returns? we should change this across all of our repositories lol

File: src/services/module_service.ts [hunk 3]
```diff
@@ -2,8 +2,13 @@ import IModule from "../interfaces/module.ts";
 import { colours, LoggerService } from "../../deps.ts";
 import DenoService from "../services/deno_service.ts";
 import NestService from "../services/nest_service.ts";
+import GitHubService from "../services/github_service.ts";
 
-const supportedUrls = ["https://deno.land/", "https://x.nest.land"];
+const supportedUrls = [
+  "https://deno.land/",
```
Comment (crookse) [inline src/services/module_service.ts:8]: can we add a comment as to why deno.land requires a trailing slash and the others don't please

File: src/services/github_service.ts (hunk 1 above)
Comment (ebebbington) [reply src/services/github_service.ts:19]: lol ;)

File: src/services/github_service.ts (hunk 2 above)
Comment (ebebbington) [reply src/services/github_service.ts:23]: I say returns?  everything we have is returns and it's just the `s` which isn't bold in the docs, still renders though

Comment (crookse) [reply src/services/github_service.ts:23]: alright

File: src/services/module_service.ts (hunk 3 above)
Comment (ebebbington) [reply src/services/module_service.ts:8]: Tbf i dont think it does, but you are right, it's probably just confusing, will change it 👍

Comment (ebebbington) [reply src/services/module_service.ts]: https://github.com/drashland/dmm/pull/91/commits/0bde1b214e9e42bbd7be9ad86e5303b3be1d2a2a

File: src/services/github_service.ts (hunk 1 above)
Comment (ebebbington) [reply src/services/github_service.ts]: https://github.com/drashland/dmm/pull/91/commits/0bde1b214e9e42bbd7be9ad86e5303b3be1d2a2a

**Review by crookse — DISMISSED** [review]
L
G
T
M

**Comment by ebebbington** [comment]
thanks for the description edit @crookse  👍

**Review by saragee3 — DISMISSED** [review]

**Review by Guergeiro — APPROVED** [review]

**Comment by saragee3** [comment]
what change prompted a re-request?

**Comment by ebebbington** [comment]
I didn't update the `--allow-net` permission to include raw.githubusercontent.com and github.com, for both the documentation and tests

So in the help message, it didnt include raw github in the install command (which would make any `dmm <command>` fail, and also we werent testing the permissions properly - there's an extra test that installs dmm with the very specific permissions then runs `dmm update` to ensure it works as expected

**Comment by ebebbington** [comment]
I just need to update the dmm docs on website too before this pr can be merged

**Comment by ebebbington** [comment]
ok docs are in staging now

**Comment by ebebbington** [comment]
ok ready for review again

**Review by crookse — APPROVED** [review]

File: src/services/module_service.ts [hunk 4]
```diff
@@ -94,6 +104,43 @@ export default class ModuleService {
...
+      importLine.indexOf("https://"),
+      importLine.indexOf(".ts") + 3, // to include the `.ts`
+    );
+    const importedVersion = ModuleService.getImportedVersionFromImportLine(
+      importLine,
+    );
+    let name = "";
+    const repoNameVersionAndFile =
+      importLine.split("https://raw.githubusercontent.com/")[1];
+    name = repoNameVersionAndFile.split("/")[1];
```
Comment (Guergeiro) [inline src/services/module_service.ts:121]: Some syntax stuff that could be added just to drop the `let` from this.
```ts
const repoNameVersionAndFile = importLine.split("https://raw.githubusercontent.com/")[1];
const name = repoNameVersionAndFile.split("/")[1] || "";
```

**Review by Guergeiro — APPROVED** [review]
Everything fine by me 😃

File: src/services/module_service.ts (hunk 4 above)
Comment (ebebbington) [reply src/services/module_service.ts:121]: Yeah you're completely right actually, i'll make the change just to enforce `const` wherever possible

Comment (ebebbington) [reply src/services/module_service.ts]: https://github.com/drashland/dmm/pull/91/commits/d24ec69d5dd3e6b44b98376df8330e698e3c8e35

**Comment by ebebbington** [comment]
tests pass, looks good

**Pull request merged**